File: 2020/hs/src/Day2.hs
module Day2 (day2) where

import Control.Monad (guard)
import Data.List.Split (splitOn)

data Policy = Policy Int Int Char

parseInput :: String -> (Policy, String)
parseInput input = (Policy lower upper char, password)
  where splitInput = splitOn ": " input
        policy = head splitInput
        password = splitInput !! 1
        splitPolicy = splitOn " " policy
        range = head splitPolicy
        char = head (splitPolicy !! 1)
        splitRange = splitOn "-" range
        lower = read $ head splitRange :: Int
        upper = read (splitRange !! 1) :: Int

isValid1 :: Policy -> String -> Bool
isValid1 (Policy lower upper char) password = lower <= count && count <= upper
  where count = length $ filter (== char) password

isValid2 :: Policy -> String -> Bool
isValid2 (Policy lower upper char) password = checkPos lower /= checkPos upper
  where checkPos pos = password !! (pos - 1) == char

day2 :: String -> IO ()
day2 input = do
  let inputs = map parseInput $ lines input
  let numValid1 = length $ filter (uncurry isValid1) inputs
  let numValid2 = length $ filter (uncurry isValid2) inputs
  guard (numValid1 == 500)
  guard (numValid2 == 313)

File: 2020/rs/src/day2.rs
use itertools::Itertools;

use crate::utils;

struct PolicyAndPassword {
    c: char,
    lower: usize,
    upper: usize,
    password: String,
}

fn parse_input(input: &str) -> PolicyAndPassword {
    let (policy, password) = input.split(": ").next_tuple().unwrap();
    let (range, c) = policy.split(' ').next_tuple().unwrap();
    let (lower, upper) = range
        .split('-')
        .map(|x| x.parse::<usize>().unwrap())
        .next_tuple()
        .unwrap();
    PolicyAndPassword {
        c: c.chars().nth(0).unwrap(),
        lower,
        upper,
        password: String::from(password),
    }
}

fn is_valid1(p: &PolicyAndPassword) -> bool {
    let count = p.password.matches(p.c).count();
    p.lower <= count && count <= p.upper
}

fn is_valid2(p: &PolicyAndPassword) -> bool {
    let a = p.password.chars().nth(p.lower - 1) == Some(p.c);
    let b = p.password.chars().nth(p.upper - 1) == Some(p.c);
    a != b
}

pub fn day2(input_path: &str) {
    let xs: Vec<PolicyAndPassword> = utils::input_lines(input_path)
        .map(|l| parse_input(&l))
        .collect();
    assert_eq!(xs.iter().filter(|x| is_valid1(x)).count(), 500);
    assert_eq!(xs.iter().filter(|x| is_valid2(x)).count(), 313);
}
